riscv-runtime/src/hash.rs

```diff
@@ -40,79 +40,100 @@ pub fn poseidon2_gl(data: &[Goldilocks; 8]) -> [Goldilocks; 8] {
-            // Note: lo/hi positions are swapped to match the Keccak machine memory layout
-            let swapped = ((val << 32) & 0xFFFFFFFF00000000) | ((val >> 32) & 0x00000000FFFFFFFF);
-            self.state[i] ^= swapped;
+        if self.next_word == Self::RATE {
+            let mut state_out = [0u32; 50];
+            keccakf(&self.state, &mut state_out);
+            self.state = state_out;
+            self.next_word = 0;
         }
+    }
+
+    pub fn update(&mut self, data: &[u8]) {
+        unsafe {
+            let (prefix, words, suffix) = data.align_to::<u32>();
+
+            for &byte in prefix {
+                self.input_buffer |= (byte as u32) << (8 * self.next_byte);
+                self.next_byte += 1;
+                if self.next_byte == 4 {
+                    self.xor_word_to_state(self.input_buffer);
+                    self.input_buffer = 0;
+                    self.next_byte = 0;
+                }
+            }
+
+            for &word in words {
+                if self.next_byte == 0 {
+                    self.xor_word_to_state(word);
+                } else {
+                    self.xor_word_to_state(self.input_buffer | (word << (8 * self.next_byte)));
+                    self.input_buffer = word >> (32 - 8 * self.next_byte);
+                }
+            }
 
-        let state_in = self.state;
-        let mut state_out = [0u64; 25];
-        keccakf(&state_in, &mut state_out);
-        self.state = state_out;
+            for &byte in suffix {
+                self.input_buffer |= (byte as u32) << (8 * self.next_byte);
+                self.next_byte += 1;
+                if self.next_byte == 4 {
+                    self.xor_word_to_state(self.input_buffer);
+                    self.input_buffer = 0;
+                    self.next_byte = 0;
+                }
+            }
+        }
     }
 
     pub fn finalize(&mut self, output: &mut [u8]) {
-        self.buffer[self.buffer_pos] = 0x01;
-        for i in self.buffer_pos + 1..self.rate {
-            self.buffer[i] = 0x00;
+        if self.next_byte > 0 {
+            self.input_buffer |= 0x01 << (8 * self.next_byte);
+            self.xor_word_to_state(self.input_buffer);
+        } else {
+            self.xor_word_to_state(0x01);
         }
-        self.buffer[self.rate - 1] ^= 0x80;
 
-        self.absorb_block();
+        while self.next_word < Self::RATE - 1 {
+            self.xor_word_to_state(0);
+        }
+        self.xor_word_to_state(0x80000000);
 
         for i in 0..4 {
-            let swapped = ((self.state[i] << 32) & 0xFFFFFFFF00000000)
-                | ((self.state[i] >> 32) & 0x00000000FFFFFFFF);
-            output[i * 8..(i + 1) * 8].copy_from_slice(&swapped.to_le_bytes());
+            output[i * 8..(i * 8 + 4)].copy_from_slice(&self.state[i * 2 + 1].to_le_bytes());
+            output[(i * 8 + 4)..(i * 8 + 8)].copy_from_slice(&self.state[i * 2].to_le_bytes());
         }
```

Comment:
This is very disturbing... basically, this code is to undo the crazy swapping done inside the keccak32_memory machine to make its interface big-endian. @leonardoalt do you know why the keccak32_memory interface is big-endian?

https://github.com/powdr-labs/powdr/blob/61510810630c5c195784f70ad68c83bd04674400/std/machines/hash/keccakf32_memory.asm#L64
https://github.com/powdr-labs/powdr/blob/61510810630c5c195784f70ad68c83bd04674400/std/machines/hash/keccakf32_memory.asm#L123

Maybe not in this PR, but if there are no other big consequences, I think keccak32_memory interface should be little-endian.

Reply:
I think the old version (no-memory) used to be little-endian, then we swapped to big-endian because it was easier to read/test, but I agree the memory version makes more sense as little-endian for memory compat.